### PR TITLE
NIFI-4099

### PR DESCRIPTION
**Pull request by lfrancke**
- Code cleanup (consistent formatting, naming etc.)
- Update of SNMP4J from 1.10.1 to 2.5.6
- Fixes restartability issue because snmpTarget was not recreated
- Updates SNMPUtils to not use Reflection (initialized from a static list of method names) to create a FlowFile

### CODE REVIEW

**Comment by lfrancke** [comment]
Both test failures are unrelated to the patch.

**Comment by trkurc** [comment]
started reviewing

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml [hunk 1]
```diff
@@ -20,8 +20,8 @@ language governing permissions and limitations under the License. -->
     <packaging>jar</packaging>
 
     <properties>
-        <snmp4j.version>1.10.1</snmp4j.version>
-        <snmp4j-agent.version>1.3.1</snmp4j-agent.version>
+        <snmp4j.version>2.5.6</snmp4j.version>
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml]: As these appear to be the sole dependency changes, I went to check the LICENSE and NOTICE of snmp4j version, to see if the nar LICENSE and NOTICE needed to be changed.  It appears that the NOTICE will need to be updated (and was probably already incorrect (see https://nifi.apache.org/licensing-guide.html), and then those changes also reflected in the nifi-assembly NOTICE

```
  (ASLv2) SNMP4J 2
    The following NOTICE information applies:
       SNMP4J
       Copyright (C) 2003-2016  Frank Fock and Jochen Katz (SNMP4J.org)

  (ASLv2) Apache log4j
    The following NOTICE information applies:
      Apache log4j
      Copyright 2007 The Apache Software Foundation
```
Also, I don't know if they're due to changes you made or not, but I don't see commons-lang or commons-io as bundled dependencies, so they could be removed from the nar NOTICE

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml [hunk 2]
```diff
@@ -20,8 +20,8 @@ language governing permissions and limitations under the License. -->
...
+        <snmp4j-agent.version>2.5.3</snmp4j-agent.version>
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml]: I noticed as I was checking LICENSE and NOTICE that there was a newer version of this (2.6.0). Curious if you considered the new version.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java [hunk 3]
```diff
@@ -47,7 +46,7 @@
      * @param snmp instance of {@link Snmp}
      * @param target instance of {@link AbstractTarget}
      */
-    public SNMPWorker(Snmp snmp, AbstractTarget target) {
+    protected SNMPWorker(Snmp snmp, AbstractTarget target) {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java]: Why the change to the access modifier? If there is code that isn't in Apache NiFi that used this class, would it break them?

Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java]: I noticed a few public->private changes as well throughout the PR. While I can't argue that the access modifier you went with would have been the best initial choice, making a change back from public can break other people's code, even if it isn't in NiFi now (in other words, the changes are a public API change).

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java [hunk 4]
```diff
@@ -48,77 +42,58 @@
 /**
  * Utility helper class that simplifies interactions with target SNMP API and NIFI API.
  */
-abstract class SNMPUtils {
+public final class SNMPUtils {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java]: Probably the right choice for this... but see my "public API comments" from before. If someone was subclassing this class...

**Comment by trkurc** [comment]
Still working through the non-style changes and working on test scenario. @lfrancke - did you have an integration test "environment" you used?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPGetter.java [hunk 5]
```diff
@@ -84,17 +83,21 @@ public ResponseEvent get() {
      * @return the list of {@link TreeEvent}
      */
     public List<TreeEvent> walk() {
-        TreeUtils treeUtils = new TreeUtils(this.snmp, new DefaultPDUFactory());
-        @SuppressWarnings("unchecked")
-        List<TreeEvent> events = treeUtils.getSubtree(this.target, this.oid);
-        return events;
+        DefaultPDUFactory pduFactory = new DefaultPDUFactory();
+
+        // BEGIN WORKAROUND - For http://oosnmp.net/pipermail/snmp4j/2016-October/005749.html
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPGetter.java]: Great comment! For maintainability, do you think it would be helpful to put a comment in the pom to revisit this code if the snmp4j version is updated?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java [hunk 6]
```diff
@@ -191,26 +164,6 @@ private static void addAttributeFromVariable(VariableBinding variableBinding, Ma
     }
 
     /**
-     * Will validate if provided name corresponds to valid SNMP property.
-     * @param name the name of the property
-     * @return 'true' if valid otherwise 'false'
-     */
-    public static boolean isValidSnmpPropertyName(String name) {
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPUtils.java]: I know this method isn't used in your improved attribute populator, but it shouldn't be removed as it is a public method in a util class.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java [hunk 7]
```diff
@@ -389,30 +391,31 @@ private AbstractTarget createSnmpTarget(ProcessContext context) {
...
 
-        if(version == SnmpConstants.version3) {
-            final String username = context.getProperty(SNMP_SECURITY_NAME).getValue();
-            final String authPassword = context.getProperty(SNMP_AUTH_PASSWORD).getValue();
-            final String privPassword = context.getProperty(SNMP_PRIV_PASSWORD).getValue();
-            final String authProtocol = context.getProperty(SNMP_AUTH_PROTOCOL).getValue();
-            final String privProtocol = context.getProperty(SNMP_PRIV_PROTOCOL).getValue();
+        AbstractTarget result;
+        if (version == SnmpConstants.version3) {
+            String username = context.getProperty(SNMP_SECURITY_NAME).getValue();
```
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java]: I'm curious, more than anything, why you removed so many of the final's throughout the code.

**Comment by lfrancke** [comment]
@trkurc Thank you for taking the time for your thorough review!

I'll go through everything you raised separately and will amend the pull request.

You are totally correct that I forgot about backwards compatibility. In that case I would add deprecation notices to the affected methods instead so they can be removed in a later version. I did a quick search just now but didn't see a policy on this.

Assuming this makes the 1.4 release could the removal happen in 1.5 or does it need to wait until 2.0 or do you have any other policy?

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml (hunk 1 above)
Comment (lfrancke) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml]: Another thing I totally forgot about/didn't know. I'll read the licensing guide and will update.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml (hunk 2 above)
Comment (lfrancke) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/pom.xml]: Released 22 June :) Good catch. I'll check and update. The agent is only used for the tests anyway so not too important. SNMP4J itself still seems to be at 2.5.6.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java (hunk 3 above)
Comment (lfrancke) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java]: In this specific case the `public` modifier didn't make any sense at all and can't have been used outside of subclasses because it is in an abstract class.

For all other cases I've already commented separately.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPGetter.java (hunk 5 above)
Comment (lfrancke) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPGetter.java]: Yes, makes sense.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java (hunk 7 above)
Comment (lfrancke) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java]: That is mostly a personal preference as I find they are mostly not needed. I know that there's coding style guidelines for this in some Hadoop related projects and most of them don't use `final` for local variables or method parameters. So I adopted this style here. But if NiFi follows another style I'm happy to keep them.

**Comment by lfrancke** [comment]
Regarding your question about the tests:

I did the same as in [NIFI-1537](https://issues.apache.org/jira/browse/NIFI-1537): I tried this with [snmpsim](https://github.com/etingof/snmpsim). That's how I stumbled across the original bug and it's how I tested this patch.

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java (hunk 7 above)
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/AbstractSNMPProcessor.java]: it isn't documented in any style guide, but I'd certainly say that there is a bias in the project towards liberal use of final for local variables.

**Comment by trkurc** [comment]
@lfrancke - you can find the versioning scheme here: https://cwiki.apache.org/confluence/display/NIFI/Version+Scheme+and+API+Compatibility

My interpretation would be wait until a major revision (i.e. 2.x) because this is an extension, and hence part of the NiFi API, and there is paragraph about issues changing method signatures later in the guide

File: nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java (hunk 3 above)
Comment (trkurc) [inline nifi-nar-bundles/nifi-snmp-bundle/nifi-snmp-processors/src/main/java/org/apache/nifi/snmp/processors/SNMPWorker.java]: totally agree with you on this case! I was guilty here of seeing only a limited section of code in my code review tool and made a bad assumption. Totally agree that while protected and public here are mostly equivalent, protected is a better choice.
